Cpp_hands_on_programming/test/test.cc

```diff
@@ -0,0 +1,133 @@
+    {"--queue", QUEUE},
+    {"--shm", SHM}
+};
+
+// Parameterized test with class fixture
+class Prog_initParamTest : public Prog_initTest, public ::testing::WithParamInterface<std::pair<const std::string, methods>> {};
+
+TEST_F(Prog_initTest, PrintInstruction) {
+    std::string expected = "This program is a file data transfer program in IPC environment, it is for IPC methods practice.\n";
+    expected += "The program shall receive an argument starts with \"--\"\n";
+    expected += "The options below are available for different file transfer usages(may require argument), --file option is mandatory:\n\n";
+    expected += "--help(-h): display all commands instruction and description(no arguments)\n";
+    expected += "--queue(-q): uses queue as IPC method(argument required)\n";
+    expected += "--pipe(-p): uses name pipe(FIFO) as IPC method(argument required)\n";
+    expected += "--shm(-s): uses shared memory buffer as IPC method(argument required)\n";
+	expected += "--file(-f): file used for read/write data(argument required)\n\n";
+    testing::internal::CaptureStdout();
+    init.printInstruction();
+    std::string actual = testing::internal::GetCapturedStdout();
+    EXPECT_THAT(expected, testing::Eq(actual));
+}
+
+TEST_P(Prog_initParamTest, CheckOptionsInfoValues) {
+    auto expected = std::make_shared<Ipc_info>("file", "argument", GetParam().second); 
+    // Simulate options and argument passed in terminal
+    std::vector<std::string> v = {GetParam().first, "--file", "file", GetParam().first, "argument"};
+    char *p[5];
+    for (int i = 0; i < v.size(); i++) {
+        p[i] = &v[i][0];
+    }
+    auto actual = init.checkOptions(5, p);
+    EXPECT_THAT(expected->getFilename(), testing::Eq(actual->getFilename()));
+    EXPECT_THAT(expected->getArgument(), testing::Eq(actual->getArgument()));
+    EXPECT_THAT(expected->getMethod(), testing::Eq(actual->getMethod()));
+}
+
+INSTANTIATE_TEST_SUITE_P(MethodSelectionVariousInputs, Prog_initParamTest, 
+testing::ValuesIn(input_label), 
+[](const testing::TestParamInfo<Prog_initParamTest::ParamType> &info) {
+    return info.param.first.substr(2);
+});
+
+std::map<std::string, std::shared_ptr<Ipc_method>> input_ipc_pointer = {
+    {"Pipe", std::make_shared<Ipc_pipe>()},
+    {"Queue", std::make_shared<Ipc_queue>()},
+    {"SHM", std::make_shared<Ipc_shm>()}
+};
+
+class GetFileSizeParamTest : public::testing::TestWithParam<std::pair<const std::string, std::shared_ptr<Ipc_method>>> {};
+
+
+TEST_P(GetFileSizeParamTest, MultipleIpcPointer) {
+    int size;;
+    size = GetParam().second->getFileSize("test/test2");
+    EXPECT_THAT(size, testing::Eq(174536)); //174536 is the size of test2 file
+}
+
+INSTANTIATE_TEST_SUITE_P(PolymorphismInput, GetFileSizeParamTest, 
+testing::ValuesIn(input_ipc_pointer), 
+[](const testing::TestParamInfo<GetFileSizeParamTest::ParamType> &info) {
+    return info.param.first;
+});
+
+TEST(PipeForkTest, SendAndReceiveData) { 
+    if (fork() > 0) {
```

Comment:
Hm,  is 'fork' works well and finished normally for you? I worry it can create two separate gtest mains, so two test outputs and maybe some more strange behaviors

Reply:
It does finish the tests. You are right though, there are two test outputs for each test and the order of the test ouput is not that organized , it probably created two gtest mains as you said, other than that the tests passed normally.

I implemented two approach for this transferring data test. The other one was creating a second cc file for receiving side, and run it in parallel with the first cc file, it did work as well. 

Is the second approach something should be done? What is the usual implementation for inter process testing ?

Reply:
Hard to say, what's the best way, maybe your 'second' approach, maybe using threads.